source/heating_model/compositional_heating.cc

```diff
@@ -0,0 +1,172 @@
+/*
+  Copyright (C) 2011 - 2017 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+
+#include <aspect/heating_model/compositional_heating.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+#include <aspect/global.h>
+
+namespace aspect
+{
+  namespace HeatingModel
+  {
+    template <int dim>
+    CompositionalHeating<dim>::CompositionalHeating ()
+    {}
+
+    template <int dim>
+    std::vector<double>
+    CompositionalHeating<dim>::
+    compute_volume_fractions( const std::vector<double> &compositional_fields) const
+    {
+      std::vector<double> volume_fractions( compositional_fields.size()+1);
+      double sum_composition = 0.0;
+
+      std::vector<double> x_comp = compositional_fields;
+      for ( unsigned int i=0; i < x_comp.size(); ++i)
+        {
+          if (field_used_in_heat_production_averaging[i] == 1)
+            {
+              // clip the compositional fields so they are between zero and one
+              // and sum them for normalization purposes
+              x_comp[i] = std::min(std::max(x_comp[i], 0.0), 1.0);
+              sum_composition += x_comp[i];
+            }
+          else
+            x_comp[i] = 0.0;
+        }
+
+      if (sum_composition >= 1.0)
+        {
+          volume_fractions[0] = 0.0;  //background material
```

Comment:
So this means if the compositions happen to add up to more than 1 you assume there is no background material? How does that work with the input file parameters? You are still expected to input a background value? This feels strange. 

Ah, but I just checked the multicomponent material model and that one works the same. So lets not worry about it for now. I would prefer to have it more explicit, whether somebody uses 1 value (expanded to all fields), N values (one per field) or N+1 values (one per field plus background field), but we should tackle that as a separate project. I opened an issue about this.